// @fabrix/spool-generics/dist/api/services/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var DataStoreGenericService_1 = require("./DataStoreGenericService");
exports.DataStoreGenericService = DataStoreGenericService_1.DataStoreGenericService;
var EmailGenericService_1 = require("./EmailGenericService");
exports.EmailGenericService = EmailGenericService_1.EmailGenericService;
var PaymentGenericService_1 = require("./PaymentGenericService");
exports.PaymentGenericService = PaymentGenericService_1.PaymentGenericService;
var FulfillmentGenericService_1 = require("./FulfillmentGenericService");
exports.FulfillmentGenericService = FulfillmentGenericService_1.FulfillmentGenericService;
var GeolocationGenericService_1 = require("./GeolocationGenericService");
exports.GeolocationGenericService = GeolocationGenericService_1.GeolocationGenericService;
var ImageGenericService_1 = require("./ImageGenericService");
exports.ImageGenericService = ImageGenericService_1.ImageGenericService;
var RenderGenericService_1 = require("./RenderGenericService");
exports.RenderGenericService = RenderGenericService_1.RenderGenericService;
var ShippingGenericService_1 = require("./ShippingGenericService");
exports.ShippingGenericService = ShippingGenericService_1.ShippingGenericService;
var TaxGenericService_1 = require("./TaxGenericService");
exports.TaxGenericService = TaxGenericService_1.TaxGenericService;
