// @fabrix/fabrix/dist/common/Service.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const events_1 = require("events");
const Generic_1 = require("./Generic");
class FabrixService extends Generic_1.FabrixGeneric {
    constructor(app) {
        super(app);
        if (!(app instanceof events_1.EventEmitter)) {
            throw new Error('The "app" argument must be of type EventEmitter');
        }
        this.app.emit(`service:${this.id}:constructed`, this);
    }
    get __() {
        if (this.app.__) {
            return this.app.__;
        }
        else {
            throw new Error('Missing spool-i18n, make sure it is included in app.main.spools');
        }
    }
    get id() {
        return this.constructor.name.replace(/(\w+)Service/, '$1').toLowerCase();
    }
    get log() {
        return this.app.log;
    }
    get config() {
        return this.app.config;
    }
    get services() {
        return this.app.services;
    }
    get models() {
        return this.app.models;
    }
}
exports.FabrixService = FabrixService;
