// @fabrix/spool-notifications/dist/api/models/ItemNotification.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const common_1 = require("@fabrix/fabrix/dist/common");
const spool_sequelize_1 = require("@fabrix/spool-sequelize");
class ItemNotification extends common_1.FabrixModel {
    static config(app, Sequelize) {
        return {
            options: {
                underscored: true
            }
        };
    }
    static schema(app, Sequelize) {
        return {
            id: {
                type: Sequelize.INTEGER,
                primaryKey: true,
                autoIncrement: true
            },
            notification_id: {
                type: Sequelize.INTEGER,
                unique: 'notification_model'
            },
            model: {
                type: Sequelize.STRING,
                unique: 'notification_model'
            },
            model_id: {
                type: Sequelize.INTEGER,
                unique: 'notification_model',
                references: null
            },
            opened: {
                type: Sequelize.BOOLEAN,
                defaultValue: false
            }
        };
    }
    static get resolver() {
        return spool_sequelize_1.SequelizeResolver;
    }
    static associate(models) {
        models.ItemNotification.belongsTo(models.Notification, {
            foreignKey: 'notification_id'
        });
    }
}
exports.ItemNotification = ItemNotification;
