// @fabrix/spool-tasks/dist/archetype/config/engine.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.engine = {
    live_mode: process.env.LIVE_MODE || true,
    auto_save: process.env.AUTO_SAVE || false,
    profile: process.env.ENGINE_PROFILE,
    crons_config: {
        auto_schedule: true,
        uptime_delay: process.env.CRON_UPTIME_DELAY || 0,
        profiles: {}
    },
    events_config: {
        auto_subscribe: true,
        profiles: {}
    },
    tasks_config: {
        auto_que: true,
        profiles: {},
        connection: {
            exchange: process.env.TASK_EXCHANGE,
            work_queue_name: process.env.TASK_WORK_QUEUE,
            interrupt_queue_name: process.env.TASK_INTERRUPT_QUEUE,
            host: process.env.TASK_RMQ_HOST,
            user: process.env.TASK_RMQ_USER,
            pass: process.env.TASK_RMQ_PASS,
            port: process.env.TASK_RMQ_PORT,
            vhost: process.env.TASK_RMQ_VHOST,
            uri: process.env.RMQ_URI,
            heartbeat: 30,
            timeout: null,
            failAfter: 60,
            retryLimit: 3,
        }
    }
};
