// @fabrix/spool-caches/dist/CachesSpool.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const extension_1 = require("@fabrix/fabrix/dist/common/spools/extension");
const caches_1 = require("./caches");
const validator_1 = require("./validator");
const config = require("./config/index");
const pkg = require("../package.json");
const api = require("./api/index");
class CachesSpool extends extension_1.ExtensionSpool {
    constructor(app) {
        super(app, {
            config: config,
            pkg: pkg,
            api: api
        });
        this.cacheInstances = {};
        this.extensions = {
            cacheInstances: {
                get: () => {
                    return this.cacheInstances;
                },
                set: (newInstances) => {
                    this.cacheInstances = newInstances;
                },
                enumerable: true,
                configurable: true
            }
        };
    }
    validate() {
        return __awaiter(this, void 0, void 0, function* () {
            if (!this.app.config.get('caches')) {
                return Promise.reject(new Error('No configuration found at config.caches!'));
            }
            return Promise.all([
                validator_1.Validator.validateCachesConfig(this.app.config.get('caches'))
            ]);
        });
    }
    configure() {
        if (this.app.config.get('caches.stores').length === 0) {
            this.app.config.set('caches.stores', [{
                    name: 'memory',
                    type: 'memory',
                    max: 500,
                    ttl: 0
                }]);
            this.app.config.set('caches.defaults', ['memory']);
        }
    }
    initialize() {
        return __awaiter(this, void 0, void 0, function* () {
            return caches_1.Caches.init(this.app);
        });
    }
    unload() {
        return __awaiter(this, void 0, void 0, function* () {
            return caches_1.Caches.unload(this.app);
        });
    }
}
exports.CachesSpool = CachesSpool;
