// @fabrix/spool-generics/dist/schemas/fulfillmentProvider/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var createOrder_1 = require("./createOrder");
exports.createOrder = createOrder_1.createOrder;
var createOrderSuccess_1 = require("./createOrderSuccess");
exports.createOrderSuccess = createOrderSuccess_1.createOrderSuccess;
var createOrders_1 = require("./createOrders");
exports.createOrders = createOrders_1.createOrders;
var createOrdersSuccess_1 = require("./createOrdersSuccess");
exports.createOrdersSuccess = createOrdersSuccess_1.createOrdersSuccess;
var destroyOrder_1 = require("./destroyOrder");
exports.destroyOrder = destroyOrder_1.destroyOrder;
var destroyOrderSuccess_1 = require("./destroyOrderSuccess");
exports.destroyOrderSuccess = destroyOrderSuccess_1.destroyOrderSuccess;
var destroyOrders_1 = require("./destroyOrders");
exports.destroyOrders = destroyOrders_1.destroyOrders;
var destroyOrdersSuccess_1 = require("./destroyOrdersSuccess");
exports.destroyOrdersSuccess = destroyOrdersSuccess_1.destroyOrdersSuccess;
var getOrder_1 = require("./getOrder");
exports.getOrder = getOrder_1.getOrder;
var getOrderSuccess_1 = require("./getOrderSuccess");
exports.getOrderSuccess = getOrderSuccess_1.getOrderSuccess;
var getOrders_1 = require("./getOrders");
exports.getOrders = getOrders_1.getOrders;
var getOrdersSuccess_1 = require("./getOrdersSuccess");
exports.getOrdersSuccess = getOrdersSuccess_1.getOrdersSuccess;
var holdOrder_1 = require("./holdOrder");
exports.holdOrder = holdOrder_1.holdOrder;
var holdOrderSuccess_1 = require("./holdOrderSuccess");
exports.holdOrderSuccess = holdOrderSuccess_1.holdOrderSuccess;
var holdOrders_1 = require("./holdOrders");
exports.holdOrders = holdOrders_1.holdOrders;
var holdOrdersSuccess_1 = require("./holdOrdersSuccess");
exports.holdOrdersSuccess = holdOrdersSuccess_1.holdOrdersSuccess;
var updateOrder_1 = require("./updateOrder");
exports.updateOrder = updateOrder_1.updateOrder;
var updateOrderSuccess_1 = require("./updateOrderSuccess");
exports.updateOrderSuccess = updateOrderSuccess_1.updateOrderSuccess;
var updateOrders_1 = require("./updateOrders");
exports.updateOrders = updateOrders_1.updateOrders;
var updateOrdersSuccess_1 = require("./updateOrdersSuccess");
exports.updateOrdersSuccess = updateOrdersSuccess_1.updateOrdersSuccess;
