// @fabrix/spool-events/dist/api/controllers/EventController.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const common_1 = require("@fabrix/fabrix/dist/common");
class EventController extends common_1.FabrixController {
    create(req, res) {
        res.json({});
    }
    findOne(req, res) {
        res.json({});
    }
    findAll(req, res) {
        const Event = this.app.models['Event'];
        const limit = Math.max(0, req.query.limit || 10);
        const offset = Math.max(0, req.query.offset || 0);
        const sort = req.query.sort || [['created_at', 'DESC']];
        const where = req.jsonCriteria(req.query.where);
        Event.findAndCountAll({
            order: sort,
            offset: offset,
            limit: limit,
            where: where
        })
            .then(events => {
            res.paginate(events.count, limit, offset, sort);
            return res.json(events.rows);
        })
            .catch(err => {
            return res.serverError(err);
        });
    }
}
exports.EventController = EventController;
