// @fabrix/spool-notifications/dist/config/spool.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.spool = {
    type: 'misc',
    provides: {
        resources: ['models', 'controllers', 'services'],
        api: {
            controllers: ['NotificationsController'],
            services: ['NotificationService'],
            models: ['User', 'Notification', 'ItemNotification']
        },
        config: ['notifications', 'routes']
    },
    lifecycle: {
        configure: {
            listen: [
                'spool:sequelize:configured',
                'spool:generics:configured',
                'spool:permissions:configured'
            ],
            emit: [
                'spool:notifications:configured'
            ]
        },
        initialize: {
            listen: [
                'spool:sequelize:initialized',
                'spool:permissions:initialized',
                'spool:generics:initialized'
            ],
            emit: [
                'spool:notifications:initialized'
            ]
        }
    }
};
