// @fabrix/spool-generics/dist/GenericsSpool.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const common_1 = require("@fabrix/fabrix/dist/common");
const validator_1 = require("./validator");
const lodash_1 = require("lodash");
const config = require("./config/index");
const pkg = require("../package.json");
const api = require("./api/index");
class GenericsSpool extends common_1.Spool {
    constructor(app) {
        super(app, {
            config: config,
            pkg: pkg,
            api: api
        });
        if (this.app.config.get('generics')) {
            const generics = this.app.config.get('generics');
            Object.keys(generics).forEach(generic => {
                if (generics[generic].hasOwnProperty('api')) {
                    lodash_1.defaultsDeep(this.api, generics[generic].api);
                }
                if (generics[generic].hasOwnProperty('config')) {
                    this.app.config.set(generic, generics[generic].config);
                }
            });
        }
    }
    validate() {
        return __awaiter(this, void 0, void 0, function* () {
            const requiredSpools = ['router'];
            const spools = Object.keys(this.app.spools);
            if (!spools.some(v => requiredSpools.indexOf(v) >= 0)) {
                return Promise.reject(new Error(`spool-generics requires spools: ${requiredSpools.join(', ')}!`));
            }
            if (!this.app.config.get('generics')) {
                return Promise.reject(new Error('No configuration found at config.generics!'));
            }
            return Promise.all([
                validator_1.Validator.validateGenericsConfig(this.app.config.get('generics'))
            ]);
        });
    }
    unload() {
        return __awaiter(this, void 0, void 0, function* () {
            return Promise.all([]);
        });
    }
}
exports.GenericsSpool = GenericsSpool;
