// @fabrix/generics-render/dist/RenderGeneric.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const MarkdownIt = require("markdown-it");
const meta = require("markdown-it-meta");
const lodash_1 = require("lodash");
const spool_generics_1 = require("@fabrix/spool-generics");
class RenderGeneric extends spool_generics_1.Generic {
    constructor(app, config) {
        super(app, config);
        this.plugins = config.plugins || [];
    }
    _init(options) {
        if (!options) {
            options = {};
        }
        options = lodash_1.defaults(options, this.config);
        const md = new MarkdownIt(options);
        md.use(meta);
        lodash_1.each(this.plugins, (plugin) => {
            if (!plugin.options) {
                plugin.options = {};
            }
            md.use(plugin.plugin, plugin.options);
        });
        return md;
    }
    render(document, options) {
        const md = this._init(options);
        const renderedDocument = md.render(document);
        const res = {
            document: renderedDocument,
            meta: md.meta || {}
        };
        return Promise.resolve(res);
    }
    renderSync(document, options) {
        const md = this._init(options);
        const renderedDocument = md.render(document);
        const res = {
            document: renderedDocument,
            meta: md.meta || {}
        };
        return res;
    }
}
exports.RenderGeneric = RenderGeneric;
