// @fabrix/spool-generics/dist/defaults/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var DataStoreDefault_1 = require("./DataStoreDefault");
exports.DataStoreDefault = DataStoreDefault_1.DataStoreDefault;
var EmailDefault_1 = require("./EmailDefault");
exports.EmailDefault = EmailDefault_1.EmailDefault;
var FulfillmentDefault_1 = require("./FulfillmentDefault");
exports.FulfillmentDefault = FulfillmentDefault_1.FulfillmentDefault;
var GeolocationDefault_1 = require("./GeolocationDefault");
exports.GeolocationDefault = GeolocationDefault_1.GeolocationDefault;
var ImageDefault_1 = require("./ImageDefault");
exports.ImageDefault = ImageDefault_1.ImageDefault;
var PaymentDefault_1 = require("./PaymentDefault");
exports.PaymentDefault = PaymentDefault_1.PaymentDefault;
var RenderDefault_1 = require("./RenderDefault");
exports.RenderDefault = RenderDefault_1.RenderDefault;
var ShippingDefault_1 = require("./ShippingDefault");
exports.ShippingDefault = ShippingDefault_1.ShippingDefault;
var TaxDefault_1 = require("./TaxDefault");
exports.TaxDefault = TaxDefault_1.TaxDefault;
