// @fabrix/spool-events/dist/api/models/EventItem.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const common_1 = require("@fabrix/fabrix/dist/common");
const spool_sequelize_1 = require("@fabrix/spool-sequelize");
class EventItem extends common_1.FabrixModel {
    static config(app, Sequelize) {
        return {
            options: {
                underscored: true
            }
        };
    }
    static schema(app, Sequelize) {
        return {
            id: {
                type: Sequelize.INTEGER,
                primaryKey: true,
                autoIncrement: true
            },
            event_id: {
                type: Sequelize.INTEGER,
                unique: 'event_object',
                notNull: true
            },
            object: {
                type: Sequelize.STRING,
                unique: 'event_object'
            },
            object_id: {
                type: Sequelize.INTEGER,
                unique: 'event_object',
                notNull: true,
                references: null
            }
        };
    }
    static get resolver() {
        return spool_sequelize_1.SequelizeResolver;
    }
    static associate(models) {
        models.EventItem.belongsTo(models.Event, {});
    }
}
exports.EventItem = EventItem;
