// @fabrix/spool-tasks/dist/TasksSpool.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const extension_1 = require("@fabrix/fabrix/dist/common/spools/extension");
const Tasks_1 = require("./Tasks");
const validator_1 = require("./validator");
const config = require("./config/index");
const pkg = require("../package.json");
const api = require("./api/index");
class TasksSpool extends extension_1.ExtensionSpool {
    constructor(app) {
        super(app, {
            config: config,
            pkg: pkg,
            api: api
        });
        this.extensions = {
            tasker: {
                get: () => {
                    return this.tasker;
                },
                set: (tasker) => {
                    throw new Error('tasker can not be set through FabrixApp, check spool-tasks instead');
                },
                enumerable: true,
                configurable: true
            }
        };
    }
    get tasker() {
        return this._tasker;
    }
    set tasker(tasker) {
        this._tasker = tasker;
    }
    validate() {
        return __awaiter(this, void 0, void 0, function* () {
            if (!this.app.config.get('tasks')) {
                return Promise.reject(new Error('No configuration found at config.tasks!'));
            }
            return Promise.all([
                validator_1.Validator.validateTasksConfig(this.app.config.get('tasks'))
            ])
                .catch(err => {
                return Promise.reject(err);
            });
        });
    }
    configure() {
        return __awaiter(this, void 0, void 0, function* () {
            return Promise.all([
                Tasks_1.Tasks.configure(this.app),
                Tasks_1.Tasks.buildTasker(this.app),
                Tasks_1.Tasks.copyDefaults(this.app)
            ])
                .catch(err => {
                return Promise.reject(err);
            });
        });
    }
    initialize() {
        return __awaiter(this, void 0, void 0, function* () {
            return Promise.all([
                Tasks_1.Tasks.addTasks(this.app)
            ])
                .catch(err => {
                return Promise.reject(err);
            });
        });
    }
    unload() {
        return __awaiter(this, void 0, void 0, function* () {
            return Promise.all([
                Tasks_1.Tasks.shutdownTasker(this.app)
            ])
                .catch(err => {
                return Promise.reject(err);
            });
        });
    }
}
exports.TasksSpool = TasksSpool;
