// @fabrix/spool-tasks/dist/Tasks.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const lodash_1 = require("lodash");
const utils_1 = require("./utils");
const Client_1 = require("./Client");
const rabbit = require('rabbot');
rabbit.nackOnError();
exports.Tasks = {
    configure: (app) => {
        return;
    },
    copyDefaults: (app) => {
        app.config.set('tasksDefaults', lodash_1.clone(app.config.get('tasks')));
        return Promise.resolve({});
    },
    buildTasker: (app) => {
        let taskerConfig = app.config.get('tasks');
        if (taskerConfig.enabled === false) {
            return Promise.resolve();
        }
        const profileName = app.config.get('tasks.profile');
        const profile = utils_1.Utils.getWorkerProfile(profileName, taskerConfig);
        taskerConfig = utils_1.Utils.configureExchangesAndQueues(profile, taskerConfig);
        app.spools.tasks.tasker = new Client_1.Client(app, rabbit, taskerConfig.exchangeName);
        return utils_1.Utils.registerTasks(profile, app, rabbit);
    },
    addTasks: (app) => {
        let taskerConfig = app.config.get('tasks');
        if (taskerConfig.enabled === false) {
            return Promise.resolve();
        }
        rabbit.configure(taskerConfig);
        return Promise.resolve();
    },
    shutdownTasker: (app) => {
        let taskerConfig = app.config.get('tasks');
        if (taskerConfig.enabled === false) {
            return Promise.resolve();
        }
        return Promise.resolve(rabbit.shutdown());
    }
};
