// @fabrix/fabrix/dist/common/spools/archetype/lib/Archetype.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const _1 = require("../../../");
const config = require("./config/index");
const pkg = require("../package.json");
const api = require("./api/index");
class Archetype extends _1.Spool {
    static get lifecycle() {
        return {};
    }
    constructor(app) {
        super(app, {
            config: config,
            api: api,
            pkg: pkg
        });
    }
    validate() {
    }
    configure() {
    }
    initialize() {
        return __awaiter(this, void 0, void 0, function* () {
        });
    }
    unload() {
        return __awaiter(this, void 0, void 0, function* () {
        });
    }
}
exports.Archetype = Archetype;
